### ISSUE TITLE
Add manual mechanism to motivate AFL to pursue a particular input

### ISSUE THREAD

**Issue by virtuald** [AFLplusplus/AFLplusplus#643]
If I know that a particular region of code is leading to something potentially problematic, it seems like it would be good to add a mechanism to manually indicate that to AFL? I imagine the API would be something like adding `__afl_this_is_interesting()` to the code -- or maybe `__afl_this_is_interesting(index)`?

Not sure how involved this might be.

**Comment by vanhauser-thc** [comment]
IMHO you would just run another afl-fuzz instance (e.g. with -S temp -o same_o_parameter -i input_dir) and in that input directory you place the file that is interesting. as it will be only one file it will be guaranteed to be fuzzed first.

you could also just instrument that parts in the code that you are interested in.

**Comment by virtuald** [comment]
The use case I'm thinking of is if I haven't yet discovered an input that has led to this code path, but prior static analysis indicates that if it ever got there then it would be *really* good to go down that path. Sure, I could figure out a way to create the input, but it seems like if I could motivate AFL to go down that path it would save some time.

**Comment by virtuald** [comment]
Another workaround would be to just add an abort() once it gets to the code path, then do the input thing that you mention. But, then I have to watch AFL until it finds the interesting input.

I might look into adding this functionality if I had some pointers for how one might add it. I've been meaning to jump into the internals but haven't had the time yet.

**Comment by vanhauser-thc** [comment]
you could have one binary where you instrument just the functions that lead to where you want to go, then automatically it would direct the fuzzing to that goal. you can use the selected instrumentation feature for that.

**Comment by virtuald** [comment]
Thinking about it a bit more, I guess a better rephrasing of what I'm looking for is a way to weight a specific location so that AFL will value it more than other paths through the executable? Disabling selected instrumentation probably doesn't help there -- after all it's possible those other locations could lead to interesting results too.

**Comment by vanhauser-thc** [comment]
ah you mean you want a guided fuzzing feature.
I fear this is something out of scope for afl++, as this would be a huge change. take a look at https://github.com/aflgo/aflgo which does exactly that.

**Comment by domenukk** [comment]
Or have a look at injon's waypoints:
https://github.com/RUB-SysSec/ijon
This can work as secondary node, to discover more paths, with afl++ as main.

**Comment by virtuald** [comment]
Thanks for the ideas!

**Comment by virtuald** [comment]
@vanhauser-thc @domenukk is there any interest in receiving IJON's changes in AFL++? The custom mutator support is pretty useful, don't really want to backport it onto IJON. :)

I've already taken the pieces of IJON that don't relate to maximizing values and ported them in AFL++ in the LTO pass, it was fairly trivial. Seems to be making some progress.

Because it directly affects the instrumentation internals and adds an extra load + xor per instrumentation block, I imagine the best way to enable/disable in AFL++ would be either with a separate compiler executable name (afl-clang-lto-ijon ? ) or by defining a macro to enable it. 

Let me know if there's interest. If you think it's out of scope for AFL++, that's fine too.

**Comment by vanhauser-thc** [comment]
@virtuald can you please give context what you are trying to achieve and why this is important for you?

I looked everywhere and cannot find a single repository or code using ijon. The burden of adding this to a target is high, so my guess is this is the reason for it.
The results are impressive when done, however I question the usefulness. as a developer you will usually have example input that reaches your code that you want to test, otherwise you would never know if it actually works. a bughunter on the other hand would not invest the time for ijon. things like crc, hashes etc. should be removed for fuzzing anyway (as it is also documented everywhere).

I am not saying no - just saying I need to hear good arguments :)

maybe @eqv and @schumilo want to chime in.

**Comment by virtuald** [comment]
My motivation is best expressed via this excerpt from their paper:

> Specifically, we focus on a particular class of challenges: we observe that current fuzzers are not able to properly explore the state space of a program beyond code coverage. For example, program executions that result in the same code coverage, but different values in the state, cannot be explored appropriately by current fuzzers.

As stated, one of the goals for their set of changes is for a tool to help expand your search once coverage has been exhausted, and that's my motivation as well.

I created a commit showing a diff between AFL 2.51b and their fork at https://github.com/virtuald/ijon/commit/dbb80e4058dc196caf5f95ab1297400ccee31a02 -- ignoring their min/max stuff, the changes to the compiler are actually fairly small.

**Comment by vanhauser-thc** [comment]
hmm it is actually quite a large change. e.g. the changes that you see in the afl-llvm-pass have to be done in LLVMInsTrim.so.cc,
  SanitizerCoverageLTO.so.cc and SanitizerCoveragePCGUARD.so.cc as well.

plus the coverage on/off functionality has to be implemented differently as this way costs too much IMHO.

and all that for something that nobody seems to be using. so I am still not convinced ...

why not running 1-2 ijon tasks in the fuzzing campaign same as cmplog, laf, eclipser etc.?
e.g. fairfuzz is also a nice fuzzer but it does not fit into afl++ as it would result in too many changes. but it can be run in parallel to afl++ (or any other variant).

**Comment by virtuald** [comment]
Fair enough.
